stdlib/option.mli

```diff
@@ -0,0 +1,78 @@
+(**************************************************************************)
+(*                                                                        *)
+(*                                 OCaml                                  *)
+(*                                                                        *)
+(*   Copyright 2018 Institut National de Recherche en Informatique et     *)
+(*     en Automatique.                                                    *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(** Option values.
+
+    @since NEXT_RELEASE *)
+
+(** {1:options Options} *)
+
+type 'a t = 'a option = None | Some of 'a
+(** The type for option values. Either [None] or [Some v]. *)
+
+val none : 'a option
+(** [none] is [None]. *)
+
+val some : 'a -> 'a option
+(** [some v] is [Some v]. *)
+
+val evict : none:'a -> 'a option -> 'a
+(** [evict ~none o] is [none] if [o] is [None] and [v] if [o] is [Some v]. *)
+
+val get : 'a option -> 'a
+(** [get o] is [v] if [o] is [Some v] and @raise Invalid_argument otherwise. *)
+
+val map : ('a -> 'b) -> 'a option -> 'b option
+(** [map f o] is [None] if [o] is [None] and [Some (f v)] is [o] is [Some v]. *)
+
+val flatten : 'a option option -> 'a option
+(** [flatten oo] is [Some v] if [oo] is [Some (Some v)] and [None]
+    otherwise. *)
+
+val iter : ('a -> unit) -> 'a option -> unit
+(** [iter f o] is [f v] if [o] is [Some v] and [()] otherwise. *)
+
+val fold : ('a -> 'b -> 'b) -> 'a option -> 'b -> 'b
+(** [fold f o acc] is [f v acc] if [o] is [Some v] and [acc] otherwise. *)
+
+val hash : ('a -> int) -> 'a option -> int
+(** [hash h o] hashes [o] with the help of [h] on [Some _] values. *)
+
+(** {1:preds Predicates and comparisons} *)
+
+val is_none : 'a option -> bool
+(** [is_none o] is [true] iff [o] is [None]. *)
+
+val is_some : 'a option -> bool
+(** [is_some o] is [true] iff [o] is [Some o]. *)
+
+val equal : ('a -> 'a -> bool) -> 'a option -> 'a option -> bool
+(** [equal eq o0 o1] is [true] iff [o0] and [o1] are both [None] or if
+   they are [Some v0] and [Some v1] and [eq v0 v1] is [true]. *)
+
+val compare : ('a -> 'a -> int) -> 'a option -> 'a option -> int
+(** [compare cmp o0 o1] is a total order on options using [cmp] to compare
+    values wrapped by [Some _]. [None] is smaller than [Some _] values. *)
+
+(** {1:convert Converting} *)
+
+val to_result : none:'e -> 'a option -> ('a, 'e) result
```

Comment:
I think it's more important to focus on the _outcome_ in this case, which is the fact that the [None] case is the error value. Thus my suggestion is to rename the labeled argument as in:

```ocaml
val to_result : error:'e -> 'a option -> ('a, 'e) result
```

**Examples**:

```ocaml
assert (Option.to_result (Some 42) ~error:"Oh no!" = Ok 42);
assert (Option.to_result None ~error:"Oh no!" = Error "Oh no!");
```

Reply:
I'm not sure I agree with this, I think using `~none` make it more clear that this is what happens on `None`.

Reply:
I found `none` slightly confusing because of the declaration of the `none` function a few lines before, but not especially so. I could go either way. That said, something like `if_none` or `on_error` or what have you strikes me as better still. That said, this is such a minor issue you should take my comment with a kilogram of salt.